### ISSUE TITLE
Error:version 2.0.2 and 2.0.1 download hurge file will segmentation fault

### ISSUE THREAD

**Issue by richardweiu** [arvidn/libtorrent#5885]
**Please provide the following information**

libtorrent version (or branch): 2.0.1 and 2.0.2

platform/architecture: centos7

compiler and compiler version: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)

please describe what symptom you see, what you would expect to see instead and
how to reproduce it.

python client run info:

```python
# download video size:4.9G
# download video by url seed
python client.py out.torrent
added torrent: opt
state updates for 1 torrents
state updates for 1 torrents
state updates for 1 torrents
state updates for 1 torrents
state updates for 1 torrents
state updates for 1 torrents
state updates for 1 torrents
Segmentation fault (core dumped)
```
gdb core dump info: [coredump.zip](https://github.com/arvidn/libtorrent/files/5836485/coredump.zip)

```bash
warning: core file may not match specified executable file.
[New LWP 1553]
[New LWP 1560]
[New LWP 1558]
[New LWP 1555]
[New LWP 1556]
[New LWP 1564]
[New LWP 1561]
[New LWP 1554]
[New LWP 1562]
[New LWP 1557]
[New LWP 1559]
[New LWP 1552]
[New LWP 1563]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `python client.py out.torrent'.
Program terminated with signal 11, Segmentation fault.
#0  0x00007f5ac6f12463 in std::_Bit_reference::operator bool (this=0x7f5ac4fafff0) at /opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_bvector.h:81
81	    { return !!(*_M_p & _M_mask); }
Missing separate debuginfos, use: debuginfo-install glibc-2.17-106.el7_2.8.x86_64 libgcc-4.8.5-4.el7.x86_64 libstdc++-4.8.5-4.el7.x86_64 python-libs-2.7.5-39.el7_2.x86_64
(gdb) list
76	    : _M_p(__x), _M_mask(__y) { }
77	
78	    _Bit_reference() _GLIBCXX_NOEXCEPT : _M_p(0), _M_mask(0) { }
79	
80	    operator bool() const _GLIBCXX_NOEXCEPT
81	    { return !!(*_M_p & _M_mask); }
82	
83	    _Bit_reference&
84	    operator=(bool __x) _GLIBCXX_NOEXCEPT
85	    {
(gdb) bt
#0  0x00007f5ac6f12463 in std::_Bit_reference::operator bool (this=0x7f5ac4fafff0) at /opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_bvector.h:81
#1  0x00007f5ac7291182 in libtorrent::hash_picker::set_block_hash (this=0x7f5ac004d750, piece=..., offset=0, h=...) at ../../src/hash_picker.cpp:356
#2  0x00007f5ac70598b9 in libtorrent::torrent::on_blocks_hashed (this=0x7f5ac000f7e0, piece=..., block_hashes=...) at ../../src/torrent.cpp:4096
#3  0x00007f5ac70589a6 in libtorrent::torrent::on_piece_verified (this=0x7f5ac000f7e0, block_hashes=..., piece=..., piece_hash=..., error=...) at ../../src/torrent.cpp:3868
#4  0x00007f5ac7074fc0 in libtorrent::torrent::<lambda(libtorrent::piece_index_t, const sha1_hash&, const libtorrent::storage_error&)>::operator()(libtorrent::piece_index_t, const libtorrent::sha1_hash &, const libtorrent::storage_error &) (__closure=0x7f5ac004da90, p=..., h=..., error=...) at ../../src/torrent.cpp:10935
#5  0x00007f5ac707d7ac in std::_Function_handler<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<160>&, const libtorrent::storage_error&), libtorrent::torrent::verify_piece(libtorrent::piece_index_t)::<lambda(libtorrent::piece_index_t, const sha1_hash&, const libtorrent::storage_error&)> >::_M_invoke(const std::_Any_data &, <unknown type in /usr/lib64/python2.7/site-packages/libtorrent.so, CU 0x15c51e1, DIE 0x17ad173>, const libtorrent::digest32<160> &, const libtorrent::storage_error &) (__functor=..., 
    __args#0=<unknown type in /usr/lib64/python2.7/site-packages/libtorrent.so, CU 0x15c51e1, DIE 0x17ad173>, __args#1=..., __args#2=...) at /opt/rh/devtoolset-7/root/usr/include/c++/7/bits/std_function.h:316
#6  0x00007f5ac72890d1 in std::function<void (libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, libtorrent::digest32<160l> const&, libtorrent::storage_error const&)>::operator()(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, libtorrent::digest32<160l> const&, libtorrent::storage_error const&) const (this=0x7f5ac00408e0, __args#0=..., __args#1=..., 
    __args#2=...) at /opt/rh/devtoolset-7/root/usr/include/c++/7/bits/std_function.h:706
#7  0x00007f5ac7287f26 in libtorrent::aux::(anonymous namespace)::caller_visitor::operator() (this=0x7f5ac4fb06b8, h=...) at ../../src/disk_io_job.cpp:67
#8  0x00007f5ac7288e5b in boost::detail::variant::invoke_visitor<libtorrent::aux::(anonymous namespace)::caller_visitor const, false>::internal_visit<std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)>&>(std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)> &, int) (this=0x7f5ac4fb05d8, operand=...) at /root/boost_1_73_0/boost/variant/variant.hpp:1028
#9  0x00007f5ac7288c33 in boost::detail::variant::visitation_impl_invoke_impl<boost::detail::variant::invoke_visitor<const libtorrent::aux::(anonymous namespace)::caller_visitor, false>, void*, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)> >(int, boost::detail::variant::invoke_visitor<libtorrent::aux::(anonymous namespace)::caller_visitor const, false> &, void *, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)> *, mpl_::true_) (visitor=..., storage=0x7f5ac00408e0) at /root/boost_1_73_0/boost/variant/detail/visitation_impl.hpp:119
#10 0x00007f5ac7288985 in boost::detail::variant::visitation_impl_invoke<boost::detail::variant::invoke_visitor<const libtorrent::aux::(anonymous namespace)::caller_visitor, false>, void*, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)>, boost::variant<std::function<void(libtorrent::disk_buffer_holder, const libtorrent::storage_error&)>, std::function<void(const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<256>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::status_t, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, const libtorrent::storage_error&)>, std::function<void()>, std::function<void(libtorrent::status_t, const libtorrent::storage_error&)>, std::function<void(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>)>, std::function<void(const libtorrent::storage_error&, libtorrent::aux::container_wrapper<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, std::vector<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, std::allocator<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void> > > >)> >::has_fallback_type_>(int, boost::detail::variant::invoke_visitor<libtorrent::aux::(anonymous namespace)::caller_visitor const, false> &, void *, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)> *, boost::variant<std::function<void(libtorrent::disk_buffer_holder, const libtorrent::storage_error&)>, std::function<void(const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<256>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::status_t, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, const libtorrent::storage_error&)>, std::function<void()>, std::function<void(libtorrent::status_t, const libtorrent::storage_error&)>, std::function<void(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>)>, std::function<void(const libtorrent::storage_error&, libtorrent::aux::container_wrapper<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, std::vector<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, std::allocator<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void> > > >)> >::has_fallback_type_, int) (internal_which=2, visitor=..., storage=0x7f5ac00408e0, t=0x0)
    at /root/boost_1_73_0/boost/variant/detail/visitation_impl.hpp:157
#11 0x00007f5ac728851d in visitation_impl<mpl_::int_<0>, boost::detail::variant::visitation_impl_step<boost::mpl::l_iter<boost::mpl::l_item<mpl_::long_<10>, std::function<void(libtorrent::disk_buffer_holder, const libtorrent::storage_error&)>, boost::mpl::l_item<mpl_::long_<9>, std::function<void(const libtorrent::storage_error&)>, boost::mpl::l_item<mpl_::long_<8>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)>, boost::mpl::l_item<mpl_::long_<7>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag>, const libtorrent::digest32<256>&, const libtorrent::storage_error&)>, boost::mpl::l_item<mpl_::long_<6>, std::function<void(libtorrent::status_t, std::basic_string<char>, const libtorrent::storage_error&)>, boost::mpl::l_item<mpl_::long_<5>, std::function<void()>, boost::mpl::l_item<mpl_::long_<4>, std::function<void(libtorrent::status_t, const libtorrent::storage_error&)>, boost::mpl::l_item<mpl_::long_<3>, std::function<void(std::basic_string<char>, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag>, const libtorrent::storage_error&)>, boost::mpl::l_item<mpl_::long_<2>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag>)>, boost::mpl::l_item<mpl_::long_<1>, std::function<void(const libtorrent::storage_error&, libtorrent::aux::container_wrapper<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag>, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag>, std::vector<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag>, std::allocator<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag> > > >)>, boost::mpl::l_end> > > > > > > > > > >, boost::mpl::l_iter<boost::mpl::l_end> >, boost::detail::variant::invoke_visitor<const libtorrent::aux::(anonymous namespace)::caller_visitor, false>, void*, boost::variant<std::function<void(libtorrent::disk_buffer_holder, const libtorrent::storage_error&)>, std::function<void(const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<160>&, const libto---Type <return> to continue, or q <return> to quit---
rrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<256>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::status_t, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, const libtorrent::storage_error&)>, std::function<void()>, std::function<void(libtorrent::status_t, const libtorrent::storage_error&)>, std::function<void(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>)>, std::function<void(const libtorrent::storage_error&, libtorrent::aux::container_wrapper<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, std::vector<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, std::allocator<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void> > > >)> >::has_fallback_type_> (no_backup_flag=..., 
    storage=0x7f5ac00408e0, visitor=..., logical_which=2, internal_which=2) at /root/boost_1_73_0/boost/variant/detail/visitation_impl.hpp:238
#12 internal_apply_visitor_impl<boost::detail::variant::invoke_visitor<libtorrent::aux::(anonymous namespace)::caller_visitor const, false>, void*> (storage=0x7f5ac00408e0, visitor=..., logical_which=2, 
    internal_which=2) at /root/boost_1_73_0/boost/variant/variant.hpp:2337
#13 internal_apply_visitor<boost::detail::variant::invoke_visitor<libtorrent::aux::(anonymous namespace)::caller_visitor const, false> > (visitor=..., this=0x7f5ac00408d8)
    at /root/boost_1_73_0/boost/variant/variant.hpp:2349
#14 boost::variant<std::function<void(libtorrent::disk_buffer_holder, const libtorrent::storage_error&)>, std::function<void(const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<256>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::status_t, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, const libtorrent::storage_error&)>, std::function<void()>, std::function<void(libtorrent::status_t, const libtorrent::storage_error&)>, std::function<void(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>)>, std::function<void(const libtorrent::storage_error&, libtorrent::aux::container_wrapper<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, std::vector<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, std::allocator<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void> > > >)> >::apply_visitor<libtorrent::aux::(anonymous namespace)::caller_visitor const>(const libtorrent::aux::(anonymous namespace)::caller_visitor &) (this=0x7f5ac00408d8, 
    visitor=...) at /root/boost_1_73_0/boost/variant/variant.hpp:2393
#15 0x00007f5ac72883a4 in boost::apply_visitor<libtorrent::aux::(anonymous namespace)::caller_visitor, boost::variant<std::function<void(libtorrent::disk_buffer_holder, const libtorrent::storage_error&)>, std::function<void(const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<256>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::status_t, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, const libtorrent::storage_error&)>, std::function<void()>, std::function<void(libtorrent::status_t, const libtorrent::storage_error&)>, std::function<void(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>)>, std::function<void(const libtorrent::storage_error&, libtorrent::aux::container_wrapper<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, std::vector<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, std::allocator<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void> > > >)> >&>(const libtorrent::aux::(anonymous namespace)::caller_visitor &, boost::variant<std::function<void(libtorrent::disk_buffer_holder, const libtorrent::storage_error&)>, std::function<void(const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<160>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>, const libtorrent::digest32<256>&, const libtorrent::storage_error&)>, std::function<void(libtorrent::status_t, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, const libtorrent::storage_error&)>, std::function<void()>, std::function<void(libtorrent::status_t, const libtorrent::storage_error&)>, std::function<void(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, const libtorrent::storage_error&)>, std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void>)>, std::function<void(const libtorrent::storage_error&, libtorrent::aux::container_wrapper<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, libtorrent::aux::strong_typedef<int, libtorrent::aux::file_index_tag, void>, std::vector<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void>, std::allocator<libtorrent::aux::strong_typedef<unsigned char, libtorrent::download_priority_tag, void> > > >)> > &) (visitor=..., visitable=...) at /root/boost_1_73_0/boost/variant/detail/apply_visitor_unary.hpp:68
#16 0x00007f5ac7288371 in libtorrent::aux::disk_io_job::call_callback (this=0x7f5ac00408a0) at ../../src/disk_io_job.cpp:133
#17 0x00007f5ac71cec6e in libtorrent::mmap_disk_io::call_job_handlers (this=0x1ee7d30) at ../../src/mmap_disk_io.cpp:1761
#18 0x00007f5ac71ce733 in libtorrent::mmap_disk_io::<lambda()>::operator()(void) const (__closure=0x7f5ac4fb0a78) at ../../src/mmap_disk_io.cpp:1729
#19 0x00007f5ac71cf94b in boost::asio::asio_handler_invoke<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()> >(libtorrent::mmap_disk_io::<lambda()> &, ...) (function=...)
    at /root/boost_1_73_0/boost/asio/handler_invoke_hook.hpp:69
#20 0x00007f5ac71cf8c6 in boost_asio_handler_invoke_helpers::invoke<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()>, libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()> >(libtorrent::mmap_disk_io::<lambda()> &, libtorrent::mmap_disk_io::<lambda()> &) (function=..., context=...)
    at /root/boost_1_73_0/boost/asio/detail/handler_invoke_helpers.hpp:37
#21 0x00007f5ac71cf890 in boost::asio::system_executor::dispatch<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()>, std::allocator<void> >(<unknown type in /usr/lib64/python2.7/site-packages/libtorrent.so, CU 0x212fb8d, DIE 0x21bf039>, const std::allocator<void> &) const (this=0x7f5ac4fb0aaf, 
    f=<unknown type in /usr/lib64/python2.7/site-packages/libtorrent.so, CU 0x212fb8d, DIE 0x21bf039>) at /root/boost_1_73_0/boost/asio/impl/system_executor.hpp:39
#22 0x00007f5ac71cf84f in boost::asio::detail::work_dispatcher<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()> >::operator()(void) (this=0x7f5ac4fb0be0)
    at /root/boost_1_73_0/boost/asio/detail/work_dispatcher.hpp:59
#23 0x00007f5ac71cf7fb in boost::asio::asio_handler_invoke<boost::asio::detail::work_dispatcher<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()> > >(boost::asio::detail::work_dispatcher<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()> > &, ...) (function=...) at /root/boost_1_73_0/boost/asio/handler_invoke_hook.hpp:69
---Type <return> to continue, or q <return> to quit---
#24 0x00007f5ac71cf776 in boost_asio_handler_invoke_helpers::invoke<boost::asio::detail::work_dispatcher<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()> >, boost::asio::detail::work_dispatcher<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()> > >(boost::asio::detail::work_dispatcher<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()> > &, boost::asio::detail::work_dispatcher<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()> > &) (function=..., context=...)
    at /root/boost_1_73_0/boost/asio/detail/handler_invoke_helpers.hpp:37
#25 0x00007f5ac71cf642 in boost::asio::detail::executor_op<boost::asio::detail::work_dispatcher<libtorrent::mmap_disk_io::add_completed_jobs_impl(libtorrent::jobqueue_t&)::<lambda()> >, std::allocator<void>, boost::asio::detail::scheduler_operation>::do_complete(void *, boost::asio::detail::scheduler_operation *, const boost::system::error_code &, std::size_t) (owner=0x1e770e0, base=0x7f5964001cc0)
    at /root/boost_1_73_0/boost/asio/detail/executor_op.hpp:70
#26 0x00007f5ac6ec0d66 in boost::asio::detail::scheduler_operation::complete (this=0x7f5964001cc0, owner=0x1e770e0, ec=..., bytes_transferred=0)
    at /root/boost_1_73_0/boost/asio/detail/scheduler_operation.hpp:40
#27 0x00007f5ac6ec29a5 in boost::asio::detail::scheduler::do_run_one (this=0x1e770e0, lock=..., this_thread=..., ec=...) at /root/boost_1_73_0/boost/asio/detail/impl/scheduler.ipp:447
#28 0x00007f5ac6ec274e in boost::asio::detail::scheduler::run (this=0x1e770e0, ec=...) at /root/boost_1_73_0/boost/asio/detail/impl/scheduler.ipp:200
#29 0x00007f5ac6ec2c1d in boost::asio::io_context::run (this=0x1e5efc0) at /root/boost_1_73_0/boost/asio/impl/io_context.ipp:63
#30 0x00007f5ac6f1ce68 in libtorrent::session::<lambda()>::operator()(void) const (__closure=0x1e5f2f8) at ../../src/session.cpp:344
#31 0x00007f5ac6f1fc47 in std::__invoke_impl<void, libtorrent::session::start(libtorrent::session_handle::session_flags_t, libtorrent::session_params&&, boost::asio::io_context*)::<lambda()> >(std::__invoke_other, <unknown type in /usr/lib64/python2.7/site-packages/libtorrent.so, CU 0xf70b6c, DIE 0x1011218>) (__f=<unknown type in /usr/lib64/python2.7/site-packages/libtorrent.so, CU 0xf70b6c, DIE 0x1011218>)
    at /opt/rh/devtoolset-7/root/usr/include/c++/7/bits/invoke.h:60
#32 0x00007f5ac6f1f9d1 in std::__invoke<libtorrent::session::start(libtorrent::session_handle::session_flags_t, libtorrent::session_params&&, boost::asio::io_context*)::<lambda()> >(<unknown type in /usr/lib64/python2.7/site-packages/libtorrent.so, CU 0xf70b6c, DIE 0x1011926>) (__fn=<unknown type in /usr/lib64/python2.7/site-packages/libtorrent.so, CU 0xf70b6c, DIE 0x1011926>)
    at /opt/rh/devtoolset-7/root/usr/include/c++/7/bits/invoke.h:95
#33 0x00007f5ac6f20000 in std::thread::_Invoker<std::tuple<libtorrent::session::start(libtorrent::session_handle::session_flags_t, libtorrent::session_params&&, boost::asio::io_context*)::<lambda()> > >::_M_invoke<0>(std::_Index_tuple<0>) (this=0x1e5f2f8) at /opt/rh/devtoolset-7/root/usr/include/c++/7/thread:234
#34 0x00007f5ac6f1ffd1 in std::thread::_Invoker<std::tuple<libtorrent::session::start(libtorrent::session_handle::session_flags_t, libtorrent::session_params&&, boost::asio::io_context*)::<lambda()> > >::operator()(void) (this=0x1e5f2f8) at /opt/rh/devtoolset-7/root/usr/include/c++/7/thread:243
#35 0x00007f5ac6f1ffb0 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<libtorrent::session::start(libtorrent::session_handle::session_flags_t, libtorrent::session_params&&, boost::asio::io_context*)::<lambda()> > > >::_M_run(void) (this=0x1e5f2f0) at /opt/rh/devtoolset-7/root/usr/include/c++/7/thread:186
#36 0x00007f5ac73f8daf in execute_native_thread_routine () from /usr/lib64/python2.7/site-packages/libtorrent.so
#37 0x00007f5aceac3dc5 in start_thread () from /lib64/libpthread.so.0
#38 0x00007f5ace0e7ced in clone () from /lib64/libc.so.6
(gdb) 
(gdb) 
(gdb) 
(gdb) info thread
  Id   Target Id         Frame 
  13   Thread 0x7f5972c9f700 (LWP 1563) 0x00007f5aceac76d5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  12   Thread 0x7f5acf2b9740 (LWP 1552) 0x00007f5ace0df413 in select () from /lib64/libc.so.6
  11   Thread 0x7f5abd7fa700 (LWP 1559) 0x00007f5aceac76d5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  10   Thread 0x7f5abe7fc700 (LWP 1557) 0x00007f5aceac76d5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  9    Thread 0x7f59734a0700 (LWP 1562) 0x00007f5aceac76d5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  8    Thread 0x7f5abffff700 (LWP 1554) 0x00007f5aceac76d5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  7    Thread 0x7f5973ca1700 (LWP 1561) 0x00007f5aceac76d5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  6    Thread 0x7f597249e700 (LWP 1564) 0x00007f5ac73b3cb1 in libtorrent::(anonymous namespace)::Rot (x=3914548225, n=17) at ../../src/sha256.cpp:58
  5    Thread 0x7f5abeffd700 (LWP 1556) 0x00007f5aceac76d5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  4    Thread 0x7f5abf7fe700 (LWP 1555) 0x00007f5aceac76d5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  3    Thread 0x7f5abdffb700 (LWP 1558) 0x00007f5aceac76d5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  2    Thread 0x7f5abcff9700 (LWP 1560) 0x00007f5ac73b3f62 in libtorrent::(anonymous namespace)::sha_compress (md=..., buf=0x7f5a184a3180 <Address 0x7f5a184a3180 out of bounds>) at ../../src/sha256.cpp:79
* 1    Thread 0x7f5ac4fb1700 (LWP 1553) 0x00007f5ac6f12463 in std::_Bit_reference::operator bool (this=0x7f5ac4fafff0) at /opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_bvector.h:81
(gdb) 
```

**Comment by arvidn** [comment]
you suggest this is triggered by downloading a large file. I assume it doesn't happen with small files then, but does it happen if you download the large file via a normal bittorrent swarm, instead of a web seed?

**Comment by richardweiu** [comment]
1. it doesn't happen with small files
2. I monitor the memory of the process when it is downloadidng, its RES and SHR increases as the download processes in 2.0.2 and 2.0.1,but in 1.0.9 version it will not
![image](https://user-images.githubusercontent.com/15622925/105116725-41fd9000-5b06-11eb-8098-7f57aedf4f84.png)
(in 2.0.2)
![image](https://user-images.githubusercontent.com/15622925/105116814-6c4f4d80-5b06-11eb-9d37-4e507afce0bb.png)
(in 1.0.9)
3. how to download large file via a normal bittorrent swarm, instead of a web seed?(I start two process, one download by web seed and share the pieces, another remove the web seed info and download by sharing, but the sharing machine will crash prematurely)

**Comment by arvidn** [comment]
> I monitor the memory of the process when it is downloadidng, its RES and SHR increases as the download processes in 2.0.2 and 2.0.1,but in 1.0.9 version it will not

This is by design. in libtorrent 2.0.x the disk cache is deferred to the kernel. This gives the operating system a much better opportunity to balance the RAM use between processes. Both by using more RAM for the torrent client (when available) and less RAM for the torrent client when not available. The virtual memory size of the process roughly corresponds to the sizes of all files currently opened (which are memory mapped) and the resident memory is the amount of physical memory the kernel has allotted for caching parts of those files. The vast majority of that resident memory is likely reclaimable by the kernel if it needs RAM for something else. (I say "likely" because some of it will be dirty pages, when downloading. It's still reclaimable but only after being flushed to disk).

> how to download large file via a normal bittorrent swarm, instead of a web seed?(I start two process, one download by web seed and share the pieces, another remove the web seed info and download by sharing, but the sharing machine will crash prematurely)

So I take it you didn't create the torrent, and you don't have the file used to create it, and even though someone else created a torrent for it, it's not seeded by bittorrent clients, but just web seeds. Something like archive.org torrents.

You could download the file with `curl` and then load it up in the torrent client. You could also test with some other file.

This is what I did to try to reproduce this:

```
./connection_tester gen-torrent -s 10000 -n 1 -t large-test.torrent
./connection_tester gen-data -t large-test.torrent -p .
```

Since `connection_tester` doesn't support v2 yet, I then re-created the torrent.

```
make_torrent large-test -2 -o large-test.torrent
```

let client test perform a full check. This exercises some of the same paths as downloading it. Build with sanitizers to ensure to catch any memory issues.

```
./client_test large-test.torrent -s large-test/
```

And I caught it, great!

**Comment by arvidn** [comment]
please test this patch: https://github.com/arvidn/libtorrent/pull/5886

This is the important part, if you want to apply it manually:
https://github.com/arvidn/libtorrent/pull/5886/files#diff-6718543712851f5e6d8903943e592385b6d99081a575e837d9988390687eb6a9R343-R344

**Comment by richardweiu** [comment]
thank you, it works!
that fantastic

**Comment by arvidn** [comment]
thanks for the report!
